### PR TITLE
Avoid holding on to DependencyContext instance

### LINKED ISSUE #4527: Avoid holding on to DependencyContext instances
They are pretty memory heavy and we should avoid keeping them around. Based on https://github.com/aspnet/Mvc/pull/4522#discussion_r60835905

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #4527

### CODE REVIEW

**Comment by pranavkm** [comment]
These were the objects and their allocations in the gen2 heap prior to this change.
image

**Comment by pranavkm** [comment]
cc @rynowak \ @Eilon

**Comment by javiercn** [comment]
Could we use application parts for this instead to calculate the metadata references? That way we solve three problems. We don't hold on to the dependency context, we configure the list of assemblies in our application at a single point and we make easier to add or remove assemblies to the compilation rather than using the callback in the options.

**Comment by pranavkm** [comment]
> Could we use application parts for this instead to calculate the metadata references?

What does that look like?

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs [hunk 1]
```diff
@@ -165,6 +151,22 @@ public CompilationResult Compile(RelativeFileInfo fileInfo, string compilationCo
...
         }
 
+        /// <summary>
+        /// Gets the <see cref="DependencyContext"/>.
+        /// </summary>
+        /// <param name="hostingEnvironment">The <see cref="IHostingEnvironment"/>.</param>
+        /// <returns>The <see cref="DependencyContext"/>.</returns>
+        protected virtual DependencyContext GetDependencyContext(IHostingEnvironment hostingEnvironment)
+        {
+            if (hostingEnvironment.ApplicationName != null)
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs]: Hmm, when is this ever null?

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs [hunk 2]
```diff
@@ -29,12 +29,12 @@ namespace Microsoft.AspNetCore.Mvc.Razor.Internal
     /// </summary>
     public class DefaultRoslynCompilationService : ICompilationService
     {
+        private readonly IHostingEnvironment _hostingEnvironment;
         private readonly IFileProvider _fileProvider;
         private readonly Action<RoslynCompilationContext> _compilationCallback;
         private readonly CSharpParseOptions _parseOptions;
         private readonly CSharpCompilationOptions _compilationOptions;
         private readonly ILogger _logger;
-        private readonly DependencyContext _dependencyContext;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs]: Will we now load the DC for **every** individual compilation? Or are the metadata references cached indefinitely?

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs (hunk 1 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs]: It's sometimes null in unit test scenarios. Although I could remove this check now that we only create it when we compile and not as part of the ctor.

File: src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs (hunk 2 above)
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs]: Nope, we cache the result of it (`_applicationReferences`) the first time we need to compile.

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Razor/Internal/DefaultRoslynCompilationService.cs]: Ok, this just seems like a good change all around then

File: test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/DefaultRoslynCompilationServiceTest.cs [hunk 3]
```diff
@@ -422,5 +413,33 @@ private DependencyContext GetDependencyContext()
...
+                      GetAccessor(viewEngineOptions),
+                      fileProviderAccessor,
+                      NullLoggerFactory.Instance)
+            {
+                _dependencyContext = dependencyContext;
+            }
+
+            private static IOptions<RazorViewEngineOptions> GetAccessor(RazorViewEngineOptions options)
+            {
+                var optionsAccessor = new Mock<IOptions<RazorViewEngineOptions>>();
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Razor.Test/Internal/DefaultRoslynCompilationServiceTest.cs]: You can just use `TestOptionsManager<>` here

**Comment by rynowak** [comment]
:shipit:

**Comment by davidfowl** [comment]
:shipit: 

@javiercn, @pranavkm and @rynowak , we should discuss adding compilation things to application parts, or using it as the seed of assemblies to find references from.

**Comment by rynowak** [comment]
We already talked about that today :laughing:

**Comment by davidfowl** [comment]
Perfect, I'd like to know what the plan is 😊

**Comment by pranavkm** [comment]
@davidfowl - https://github.com/aspnet/Mvc/issues/4498 tracks it. We'll introduce a new application feature to represent compilation assets and populate it at the same time the application manager is created.

**Pull request closed without merging**